### PR TITLE
feat: Select and Multiselect V2

### LINKED ISSUE #7259: MultiSelect widget cannot be dragged by its body

### LINKED ISSUE #10297: [Enhancement]: Select and Multi-select V2 to handle server side filtering
Currently, select and multi-select widgets are not able to support server side filtering because we are not storing the labels of values selected from the server side. 

This causes the multi-select and select not usable. We need to store the label and the value of the selected options.

PRD

# Select widgets v2

## Objective

To support server side filtering in select and multi-select widgets, we need to launch v2 of the widgets where we save the label selected along with value. Today we only save the value and rely on the data to fetch the label, but this is not possible when server sider filtering is turned on and option is not present in the filtered list

## Success Metrics *

| Goal | Metric |
| --- | --- |
| Support server-side filtering in select widgets | % of dropdowns with server-side filtering switch on. (Need analytics event) |

Requirements

| Requirement | User story | Comments |
| --- | --- | --- |
| App viewer should view the label of the selected option in the input for select and multi-select | As a user of the app, I want to see the option I have selected get reflected in the input for select widgets. |  |
| If they search for the same option again in the dropdown they should see the option selected to be highlighted | As a user of the app, I want to see the option I have selected get highlighted if I search for it. |  |
| Multi-select widget should have a search bar inside the dropdown like select. Provide the filterable option here for the developer to enable. | As a developer, I want to allow users to search for options in a multi-select widget. | We need a filterable toggle similar to select widget in multi-select for the search inside the dropdown |
| Developer will now configure the default option(s) instead of value both for client and server side. Default option needs both “value” and “label”. Will be applicable for any new select/multi-select widget. Multi-select will have default options, where user is supposed to add array of objects. | As a developer, I want to specify the value and label for the default option(s) to show if the data is not available on client side. | While we can keep the client side to be similar to current behavior, this will be inconsistent for the user as if they enable server side they have to add a default label. 

This will impact a lot of areas though - 1. Cypress tests 2. Documentation 3. Videos |

## Documentation

- [Add] Multi-select filtering toggle 
- [Update] Default option for select and Default options for multi-select

## Out of Scope

- Changes to tree and multi-tree select widget

## Developer Handoff Document in Figma
Will need design discussion. Most things are re-usable.

## RACI Matrix

| Responsible |  |
| --- | --- |
| Accountable |  |
| Consulted |  |
| Informed |  |

### PR DESCRIPTION
**Pull request by Tooluloope**
## Description

Select and MultiSelect Version 2
Fixes Server Side issues

Fixes #10297

## Type of change

> Please delete options that are not relevant.

- Bugfix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: fix/Select-and-MultiSelect-V2 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.93 **(-0.13)** | 36.38 **(-0.24)** | 34.61 **(-0.17)** | 55.34 **(-0.12)**
 :red_circle: | app/client/src/icons/WidgetIcons.tsx | 59.34 **(-0.89)** | 100 **(0)** | 15.91 **(-0.76)** | 59.34 **(-0.89)**
 :red_circle: | app/client/src/sagas/SnipingModeSagas.ts | 20 **(-0.59)** | 3.13 **(-0.1)** | 40 **(0)** | 20.79 **(-0.64)**
 :red_circle: | app/client/src/utils/DSLMigrations.ts | 73.16 **(-0.14)** | 69.39 **(-0.17)** | 71.13 **(0)** | 72.96 **(-0.14)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :sparkles: :new: | **app/client/src/widgets/MultiSelectWidgetV2/index.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/widgets/MultiSelectWidgetV2/component/index.styled.tsx** | **45** | **28.13** | **0** | **51.52**
 :sparkles: :new: | **app/client/src/widgets/MultiSelectWidgetV2/component/index.tsx** | **12.77** | **0** | **0** | **14.46**
 :sparkles: :new: | **app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx** | **66.67** | **0** | **60** | **67.65**
 :sparkles: :new: | **app/client/src/widgets/SelectWidget/index.ts** | **100** | **100** | **100** | **100**
 :sparkles: :new: | **app/client/src/widgets/SelectWidget/component/index.styled.tsx** | **43.75** | **21.88** | **0** | **48.28**
 :sparkles: :new: | **app/client/src/widgets/SelectWidget/component/index.tsx** | **25** | **0** | **7.14** | **25.4**
 :sparkles: :new: | **app/client/src/widgets/SelectWidget/widget/index.tsx** | **49.02** | **0** | **46.15** | **51.06**
 :green_circle: | app/client/src/workers/validations.ts | 76.79 **(0)** | 69.49 **(0.12)** | 67.65 **(0)** | 78.59 **(0)**</details>

### CODE REVIEW

**Review by sbalaji1192 — CHANGES_REQUESTED** [review]

File: app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+import { EvaluationSubstitutionType } from "entities/DataTree/dataTreeFactory";
+import MultiSelectComponent from "../component";
+import {
+  DefaultValueType,
+  LabelValueType,
+} from "rc-select/lib/interface/generator";
+import { Layers } from "constants/Layers";
+import { MinimumPopupRows, GRID_DENSITY_MIGRATION_V1 } from "widgets/constants";
+
+class MultiSelectWidget extends BaseWidget<
```
[the remark this answers is not in the document]
Comment (Tooluloope) [reply app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx:16]: For now no.
I'm going to continue that in this issue https://github.com/appsmithorg/appsmith/issues/10575

File: app/client/src/utils/autocomplete/EntityDefinitions.ts [hunk 2]
```diff
@@ -160,6 +160,38 @@ export const entityDefinitions: Record<string, unknown> = {
     isDisabled: "bool",
     options: "[dropdownOption]",
   },
+  SELECT_WIDGET: {
+    "!doc":
+      "Select is used to capture user input/s from a specified list of permitted inputs. A Select can capture a single choice as well as multiple choices",
```
Comment (sbalaji1192) [inline app/client/src/utils/autocomplete/EntityDefinitions.ts:165]: Select widget can only capture single option, right?

**Review by sbalaji1192 — APPROVED** [review]

**Comment by vivekverma2312** [comment]
@Tooluloope After setting default option the search functionality stops working for select widget.
https://www.loom.com/share/7360a162ea7a438290e8dbb71e893aa5

**Comment by vivekverma2312** [comment]
@Tooluloope Tracking all the issues here.
1. Observed scroll issue on select widget
https://www.loom.com/share/598d283bdf4c4b93bdf7d250f571834f

2. Two options are showing as highlighted. After we close the drop-down options and open still it's displayed
https://www.loom.com/share/d68fdc23f61f42debb7b64975a7fdeba

3. Not able to see the drop-down options after we filter and navigate to query page and come back. After we clear the filter result also it do not populate. 
https://www.loom.com/share/24508625ada74cf291d7d82f37c68251

4. After setting default the search functionality stops working https://www.loom.com/share/7360a162ea7a438290e8dbb71e893aa5

**Comment by Tooluloope** [comment]
@vivekverma2312 Fixed all

**Comment by vivekverma2312** [comment]
@Tooluloope 
1. Issue2 still exist for multi-select drop-down
2. Issue 3 is still reproducible (Now I do not have option to clear the filter)
3. Drop-down size of select and multi-select is not same.
Screenshot 2022-01-28 at 12 26 44 PM
Screenshot 2022-01-28 at 12 26 38 PM

**Comment by Tooluloope** [comment]
Hi @vivekverma2312 
Concerning issue issue2, Unlike Select that has a single Option , MultiSelect has multiple options, Also the Library doesn't expose API to resume focus of the dropdown from the previously Select option. And I think it has a better UX like this compared to resuming from the Top
cc: @dilippitchika 

Issue 3 is now fixed.

Lastly, I've made the dropdown accommodate Equal number of items (8 options)

**Comment by vivekverma2312** [comment]
@Tooluloope Observed that widget icon is not coming on Entity Explorer. Rest of the issues are tested and working fine
Screenshot 2022-01-31 at 12 26 13 PM
.

**Comment by Tooluloope** [comment]
@vivekverma2312 Could you check again.
Just fixed this

**Comment by vivekverma2312** [comment]
Thanks @Tooluloope Tested and working fine now.

File: app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+      selectedOptionLabels: `{{ this.selectedOptions ? this.selectedOptions.map((o) => o.label ) : [] }}`,
+      selectedOptionValues: `{{ this.selectedOptions ? this.selectedOptions.map((o) => o.value ) : [] }}`,
+      isValid: `{{this.isRequired ? !!this.selectedOptionValues && this.selectedOptionValues.length > 0 : true}}`,
+    };
+  }
+
+  static getDefaultPropertiesMap(): Record<string, string> {
+    return {
+      selectedOptions: "defaultOptionValue",
+      filterText: "",
```
Comment (Rishabh-Rathod) [inline app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx:316]: Hey @Tooluloope, 

Could you help me understand why do we have  `filterText: "",` here. 
if `filterText` doesn't have a default property then according to my knowledge. It is fine to not add it in `getDefaultPropertiesMap`.

Comment (Tooluloope) [reply app/client/src/widgets/MultiSelectWidgetV2/widget/index.tsx:316]: Yes, @Rishabh-Rathod its okay to remove it from the Map

**Pull request merged**